File: atd/priority.py
"""优先级双模式：档位（levels）与加权评分（urgency）。"""
from __future__ import annotations

from datetime import date, datetime, timedelta

from . import config
from .model import Task


def level_rank(task: Task, levels: list[str]) -> int:
    """档位序号，越大越优先；无档位视为最低。"""
    try:
        return levels.index(task.priority)
    except ValueError:
        return -1


def urgency(task: Task, *, cfg: dict | None = None, now: datetime | None = None) -> float:
    """TaskWarrior 式加权评分，系数取自 config 的 priority.urgency 段。"""
    cfg = cfg or config.load()
    u = cfg["priority"]["urgency"]
    now = now or datetime.now()
    today = now.date()
    score = 0.0
    if task.due is not None:
        dd = task.due.date()
        if task.status == "todo":
            if dd < today:
                overdue_days = min((today - dd).days, 7)
                score += u["overdue"] * overdue_days / 7.0
            elif dd == today:
                score += u["due_today"]
            else:
                ahead = (dd - today).days
                if ahead <= 7:
                    score += u["due_week_decay"] * (1 - ahead / 7.0)
    levels = config.levels(cfg)
    if task.priority and task.priority in levels:
        score += u["per_level"] * (levels.index(task.priority) + 1) / len(levels)
    if task.entry:
        try:
            entry = datetime.fromisoformat(task.entry)
            if entry.tzinfo is not None:
                entry = entry.astimezone().replace(tzinfo=None)
            age_days = max((now - entry).days, 0)
            score += min(age_days * u["age_per_day"], u["age_cap"])
        except ValueError:
            pass
    if task.status == "waiting":
        score -= u["waiting_penalty"]
    return round(score, 3)


def sort_key(task: Task, *, mode: str, cfg: dict | None = None, levels: list[str] | None = None,
             now: datetime | None = None) -> tuple:
    """统一排序键：逾期最先、今天次之；档位模式按日期升序再按档位，
    urgency 模式按分数降序再按日期。"""
    cfg = cfg or config.load()
    levels = levels or config.levels(cfg)
    now = now or datetime.now()
    today = now.date()
    if task.due is not None:
        dd = task.due.date()
        if task.status == "todo" and dd < today:
            bucket = 0  # 逾期（日期升序 = 超得最久的在前）
        elif dd == today:
            bucket = 1
        else:
            bucket = 2
        due_ord = (dd - date(2000, 1, 1)).days
    else:
        bucket, due_ord = 3, 99999
    if mode == "urgency":
        return (bucket, -urgency(task, cfg=cfg, now=now), due_ord, task.title)
    return (bucket, due_ord, -level_rank(task, levels), task.title)


def sort_tasks(tasks: list[Task], *, mode: str | None = None, cfg: dict | None = None,
               now: datetime | None = None) -> list[Task]:
    cfg = cfg or config.load()
    mode = mode or config.priority_mode(cfg)
    levels = config.levels(cfg)
    return sorted(tasks, key=lambda t: sort_key(t, mode=mode, cfg=cfg, levels=levels, now=now))

File: src/core/priority.ts
import type { Config, Task } from "../contracts.js";
import { configLevels, loadConfig, priorityMode } from "./config.js";
import { localDate, localNow } from "./task.js";

const dateOrdinal = (date: string): number => Math.floor(Date.parse(`${date}T00:00:00Z`) / 86_400_000);
export const levelRank = (task: Task, levels: string[]): number => {
  const rank = levels.indexOf(task.priority ?? "");
  return rank < 0 ? -1 : rank;
};

export const urgency = (task: Task, config: Config, now = localNow()): number => {
  const coefficients = config.priority.urgency;
  const today = now.slice(0, 10);
  let score = 0;
  if (task.due && task.status === "todo") {
    const due = localDate(task.due);
    const delta = dateOrdinal(due) - dateOrdinal(today);
    if (delta < 0) score += coefficients.overdue * Math.min(-delta, 7) / 7;
    else if (delta === 0) score += coefficients.due_today;
    else if (delta <= 7) score += coefficients.due_week_decay * (1 - delta / 7);
  }
  const levels = configLevels(config);
  if (task.priority && levels.includes(task.priority)) score += coefficients.per_level * (levels.indexOf(task.priority) + 1) / levels.length;
  if (task.entry) {
    const entryMs = Date.parse(task.entry);
    const nowMs = Date.parse(`${now}:00Z`);
    if (Number.isFinite(entryMs) && Number.isFinite(nowMs)) score += Math.min(Math.max(Math.floor((nowMs - entryMs) / 86_400_000), 0) * coefficients.age_per_day, coefficients.age_cap);
  }
  if (task.status === "waiting") score -= coefficients.waiting_penalty;
  return Math.round(score * 1000) / 1000;
};

export type SortMode = "levels" | "urgency";
export const sortKey = (task: Task, mode: SortMode, config: Config, now = localNow()): [number, number, number, string] => {
  const today = now.slice(0, 10);
  const due = task.due ? localDate(task.due) : undefined;
  const bucket = due ? task.status === "todo" && due < today ? 0 : due === today ? 1 : 2 : 3;
  const dueOrd = due ? dateOrdinal(due) - dateOrdinal("2000-01-01") : 99_999;
  return mode === "urgency" ? [bucket, -urgency(task, config, now), dueOrd, task.title] : [bucket, dueOrd, -levelRank(task, configLevels(config)), task.title];
};

export const sortTasks = async (tasks: Task[], mode: SortMode | undefined, config?: Config, now?: string): Promise<Task[]> => {
  const cfg = config ?? await loadConfig();
  const selected = mode ?? priorityMode(cfg);
  return [...tasks].sort((a, b) => compareTuple(sortKey(a, selected, cfg, now), sortKey(b, selected, cfg, now)));
};

export const compareTuple = (a: [number, number, number, string], b: [number, number, number, string]): number => {
  for (let i = 0; i < a.length; i += 1) {
    const av = a[i]!;
    const bv = b[i]!;
    if (av === bv) continue;
    if (typeof av === "string" && typeof bv === "string") return av.localeCompare(bv);
    return Number(av) - Number(bv);
  }
  return 0;
};
